watch/watch_test.go

```diff
@@ -0,0 +1,106 @@
+package watch
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+	"time"
+
+	devtest "github.com/puma/puma-dev/dev/devtest"
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	tmpDir        = filepath.Join(devtest.ProjectRoot, "tmp")
+	tmpFilename   = "restart.txt"
+	tmpRestartTxt = filepath.Join(tmpDir, tmpFilename)
+)
+
+type Notice struct{}
+
+func TestMain(m *testing.M) {
+	if err := fileCreateAll(tmpRestartTxt); err != nil {
+		panic(err)
+	}
+
+	result := m.Run()
+
+	os.RemoveAll(tmpDir)
+
+	os.Exit(result)
+}
+
+func TestWatch_ExpectTimeout(t *testing.T) {
+	watchTriggered := watchTmpFileWithTimeout(t, func() {})
+
+	assert.False(t, watchTriggered)
+}
+
+func TestWatch_ExpectTouchSignalAfterModify(t *testing.T) {
+	watchTriggered := watchTmpFileWithTimeout(t, func() {
+		touchFile(t, tmpRestartTxt)
+	})
+
+	assert.True(t, watchTriggered)
+}
+
+func fileCreateAll(fullPath string) error {
+	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
+		return err
+	}
+
+	f, err := os.Create(fullPath)
+
+	if err != nil {
+		return err
+	}
+
+	f.Close()
+
+	return nil
+}
+
+func touchFile(t *testing.T, fullPath string) {
+	if err := exec.Command("sh", "-c", fmt.Sprintf("touch %s", fullPath)).Run(); err != nil {
+		t.FailNow()
```

Comment:
if you call t.FailNow() inside a goroutine, it can fail with "panic: Log in goroutine after <method> has completed"

probably safer to go with `assert.Fail(t, err.Error()`

Reply:
lol. called it. https://travis-ci.org/puma/puma-dev/jobs/653816457?utm_medium=notification&utm_source=github_status